compiler/compiler.go

```diff
@@ -314,11 +316,72 @@ func WritePkgCode(pkg *Archive, dceSelection map[*Decl]struct{}, gls goLinknameS
 		if _, err := w.Write(d.DeclCode); err != nil {
 			return err
 		}
+		for _, linkName := range d.IMethodLinkingNames {
+			if gls.IsImplementation(linkName) {
+				// This decl is referenced by a go:linkname directive, expose it to external
+				// callers via $linkname object (declared in prelude). We are not using
+				// $pkg to avoid clashes with exported symbols.
+				var code string
+				n := strings.LastIndexByte(linkName.Name, '.')
+				code = fmt.Sprintf("\t$linknames[%q] = function(s) { return s.%v(...[...arguments].slice(1))};\n", linkName.String(), linkName.Name[n+1:])
+				if _, err := w.Write(removeWhitespace([]byte(code), minify)); err != nil {
+					return err
+				}
+			}
+		}
 		if gls.IsImplementation(d.LinkingName) {
 			// This decl is referenced by a go:linkname directive, expose it to external
 			// callers via $linkname object (declared in prelude). We are not using
 			// $pkg to avoid clashes with exported symbols.
-			code := fmt.Sprintf("\t$linknames[%q] = %s;\n", d.LinkingName.String(), d.Vars[0])
+			var code string
+			if recv, method, ok := d.LinkingName.IsMethod(); ok {
+				if strings.HasPrefix(recv, "*") {
+					code = fmt.Sprintf(`$linknames[%q] = function(v) {
+	var T = %v;
+	var r = v;
+	var ptrType = $ptrType(T);
+	if (v.constructor != ptrType) {
+		switch (T.kind) {
+		case $kindStruct:
+			r = $pointerOfStructConversion(v, ptrType);
+			break;
+		case $kindArray:
+			r = new ptrType(v);
+			break;
+		default:
+			r = new ptrType(v.$get,v.$set,v.$target);
+		}
+	}
+	return r.%v(...[...arguments].slice(1));
+};
+`, d.LinkingName.String(), recv[1:], method)
```

Comment:
Why not use the existing `$methodExpr` helper?

I may be mistaken, but semantically these two programs should be exactly equivalent:

Example 1:
```go
package main

type X struct{}

func (x X) Foo(i int) { println(i) }

func main() {
	x := X{}
	X.Foo(x, 1)  // Note the special method calling syntax here.
}
```

Example 2:
```go
package main

import (
	_ "unsafe"
)

type X struct{}

func (x X) Foo(i int) { println(i) }

//go:linkname X_Foo main.X.Foo
func X_Foo(x X, i int)

func main() {
	x := X{}
	X_Foo(x, 1)
}
```

We already support example 1 syntax using the `$methodExpr()` helper, so I think it should be reusable for the example 2 with the edits I suggested above.

Reply:
$methodToFunction must support type convert for compatible go compiler

eample  the main.name must convert to unexport reflect.name by call name_isExported
```
import (
    _ "unsafe"
    _ "reflect"
)
// compatible to reflect.name
type name struct {
	bytes *byte
}
//go:linkname name_isExported reflect.name.isExported
func name_isExported(name) bool
```

Reply:
I see, I missed that. I guess this is necessary to reattach different methods to the object, in case the linked method calls other methods on that type. In that sense GopherJS diverges from native Go, where the appropriate method table is determined and fixed at compile time, rather than by being pointed at by the object itself.

Then having a different helper makes sense. Although I would note that the same problem remains with arguments of the method. If we are implicitly changing types for those, their method calls may fail as well. Not sure how big of a problem that is.